Grab stats and libwebrtc logs from WebInspector directly

youennf opened the pull request:
#### 7533deb1da033f11212d061fad58013f8dc401fb
<pre>
Grab stats and libwebrtc logs from WebInspector directly
<a href="https://rdar.apple.com/122803271">rdar://122803271</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=269196">https://bugs.webkit.org/show_bug.cgi?id=269196</a>

Reviewed by Eric Carlson and Devin Rousso.

To help web developpers, it is convenient to get access to WebRTC stats and logs.
We add a Web Inspector new function that allows to pass a JS callback to get those logs.
This can be extended in the future to rtc event logs as well.
This new function is called gatherRTCLogs and is only available in Web Inspector command line.

* LayoutTests/http/tests/inspector/gatherWebInspectorRTCLogs-expected.txt: Added.
* LayoutTests/http/tests/inspector/gatherWebInspectorRTCLogs.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.exp:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/mediastream/PeerConnectionBackend.h:
(WebCore::PeerConnectionBackend::startGatheringStatLogs):
(WebCore::PeerConnectionBackend::stopGatheringStatLogs):
* Source/WebCore/Modules/mediastream/RTCController.cpp:
(WebCore::RTCController::RTCController):
(WebCore::RTCController::~RTCController):
(WebCore::RTCController::add):
(WebCore::RTCController::startGatheringLogs):
(WebCore::RTCController::stopGatheringLogs):
(WebCore::RTCController::startGatheringStatLogs):
(WebCore::RTCController::stopLoggingLibWebRTCLogs):
* Source/WebCore/Modules/mediastream/RTCController.h:
* Source/WebCore/Modules/mediastream/RTCPeerConnection.cpp:
(WebCore::RTCPeerConnection::startGatheringStatLogs):
(WebCore::RTCPeerConnection::stopGatheringStatLogs):
* Source/WebCore/Modules/mediastream/RTCPeerConnection.h:
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.cpp:
(WebCore::RTCStatsLogger::toJSONString const):
(WebCore::LibWebRTCMediaEndpoint::OnStatsDelivered):
(WebCore::LibWebRTCMediaEndpoint::statsLogInterval const):
(WebCore::LibWebRTCMediaEndpoint::startRTCLogs):
(WebCore::LibWebRTCMediaEndpoint::stopRTCLogs):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.h:
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCPeerConnectionBackend.cpp:
(WebCore::LibWebRTCPeerConnectionBackend::startGatheringStatLogs):
(WebCore::LibWebRTCPeerConnectionBackend::stopGatheringStatLogs):
(WebCore::LibWebRTCPeerConnectionBackend::provideStatLogs):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCPeerConnectionBackend.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::startGatheringRTCLogs):
(WebCore::Document::stopGatheringRTCLogs):
* Source/WebCore/dom/Document.h:
* Source/WebCore/inspector/CommandLineAPIHost.cpp:
(WebCore::CommandLineAPIHost::gatherRTCLogs):
* Source/WebCore/inspector/CommandLineAPIHost.h:
* Source/WebCore/inspector/CommandLineAPIHost.idl:
* Source/WebCore/inspector/CommandLineAPIModuleSource.js:
(injectModule):
* Source/WebCore/inspector/RTCLogsCallback.h: Added.
* Source/WebCore/inspector/RTCLogsCallback.idl: Added.
* Source/WebCore/platform/SourcesLibWebRTC.txt:
* Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCLogSink.cpp: Added.
(WebCore::LibWebRTCLogSink::LibWebRTCLogSink):
(WebCore::LibWebRTCLogSink::~LibWebRTCLogSink):
(WebCore::toWebRTCLogLevel):
(WebCore::LibWebRTCLogSink::logMessage):
(WebCore::LibWebRTCLogSink::start):
(WebCore::LibWebRTCLogSink::stop):
* Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCLogSink.h: Added.
* Source/WebInspectorUI/UserInterface/Controllers/JavaScriptRuntimeCompletionProvider.js:
(WI.JavaScriptRuntimeCompletionProvider.get _commandLineAPIKeys):

Canonical link: <a href="https://commits.webkit.org/274946@main">https://commits.webkit.org/274946@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6fdacddd42bf310f99d34b180627159f1e865f2d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40461 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19473 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42839 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43011 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36550 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42768 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22428 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16802 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33569 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41035 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16425 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34881 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14153 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14220 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35853 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44287 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36686 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36176 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39923 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15276 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12515 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38224 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16895 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/35136 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9070 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16945 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16539 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/JavaScriptCore/inspector/InjectedScriptSource.js:
```diff
@@ -1683,6 +1683,8 @@ CommandLineAPI.methods["timeStamp"] = function() { return inspectedGlobalObject.
 CommandLineAPI.methods["trace"] = function() { return inspectedGlobalObject.console.trace(...createIterableWithoutPrototypeFromArguments(arguments)); };
 CommandLineAPI.methods["warn"] = function() { return inspectedGlobalObject.console.warn(...createIterableWithoutPrototypeFromArguments(arguments)); };
 
+CommandLineAPI.methods["gatherRTCLogs"] = function() { return inspectedGlobalObject.@gatherRTCLogs(...createIterableWithoutPrototypeFromArguments(arguments)); };
```
you may be better off moving this to [Source/WebCore/inspector/CommandLineAPIModuleSource.js](https://github.com/WebKit/WebKit/blob/main/Source/WebCore/inspector/CommandLineAPIModuleSource.js) as then it will only exist for pages (i.e. not for things like `JSContext`) and using the `CommandLineAPIHost` to bridge between JS and C++

youennf replied:
I see, will do.

youennf commented on Source/WebCore/DerivedSources-output.xcfilelist:
```diff
@@ -101,6 +101,8 @@ $(BUILT_PRODUCTS_DIR)/DerivedSources/WebCore/JSAnimationPlaybackEventInit.cpp
 $(BUILT_PRODUCTS_DIR)/DerivedSources/WebCore/JSAnimationTimeline.cpp
 $(BUILT_PRODUCTS_DIR)/DerivedSources/WebCore/JSAnimationTimeline.h
+$(BUILT_PRODUCTS_DIR)/DerivedSources/WebCore/JSAnyCallback.cpp
+$(BUILT_PRODUCTS_DIR)/DerivedSources/WebCore/JSAnyCallback.h
```
Will need to remove these, not sure why there are still there.

eric-carlson reviewed (approved):
r+ given Devin's previous reviews

dcrousso commented on Source/WebCore/inspector/CommandLineAPIModuleSource.js:
```diff
@@ -189,6 +189,8 @@ if (inspectedGlobalObject.document && inspectedGlobalObject.Node) {
     };
 }
 
+CommandLineAPI.methods["gatherRTCLogs"] = function() { return CommandLineAPIHost.gatherRTCLogs(arguments[0]); };
```
we may want to conditionalize this just in case `ENABLE_WEB_RTC` is false

so either this
```js
if (CommandLineAPIHost.gatherRTCLogs) {
    CommandLineAPI.methods["gatherRTCLogs"] = function(callback) {
        return CommandLineAPIHost.gatherRTCLogs(callback);
    };
}
```
or this
```js
CommandLineAPI.methods["gatherRTCLogs"] = function(callback) {
    return CommandLineAPIHost.gatherRTCLogs?.(callback);
};
```

youennf replied:
Done

dcrousso reviewed (approved):
r=me for the Web Inspector bits.  thanks for iterating on this with me :)

youennf committed:
Grab stats and libwebrtc logs from WebInspector directly
rdar://122803271
https://bugs.webkit.org/show_bug.cgi?id=269196

Reviewed by Eric Carlson and Devin Rousso.

To help web developpers, it is convenient to get access to WebRTC stats and logs.
We add a Web Inspector new function that allows to pass a JS callback to get those logs.
This can be extended in the future to rtc event logs as well.
This new function is called gatherRTCLogs and is only available in Web Inspector command line.

* LayoutTests/http/tests/inspector/gatherWebInspectorRTCLogs-expected.txt: Added.
* LayoutTests/http/tests/inspector/gatherWebInspectorRTCLogs.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.exp:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Modules/mediastream/PeerConnectionBackend.h:
(WebCore::PeerConnectionBackend::startGatheringStatLogs):
(WebCore::PeerConnectionBackend::stopGatheringStatLogs):
* Source/WebCore/Modules/mediastream/RTCController.cpp:
(WebCore::RTCController::RTCController):
(WebCore::RTCController::~RTCController):
(WebCore::RTCController::add):
(WebCore::RTCController::startGatheringLogs):
(WebCore::RTCController::stopGatheringLogs):
(WebCore::RTCController::startGatheringStatLogs):
(WebCore::RTCController::stopLoggingLibWebRTCLogs):
* Source/WebCore/Modules/mediastream/RTCController.h:
* Source/WebCore/Modules/mediastream/RTCPeerConnection.cpp:
(WebCore::RTCPeerConnection::startGatheringStatLogs):
(WebCore::RTCPeerConnection::stopGatheringStatLogs):
* Source/WebCore/Modules/mediastream/RTCPeerConnection.h:
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.cpp:
(WebCore::RTCStatsLogger::toJSONString const):
(WebCore::LibWebRTCMediaEndpoint::OnStatsDelivered):
(WebCore::LibWebRTCMediaEndpoint::statsLogInterval const):
(WebCore::LibWebRTCMediaEndpoint::startRTCLogs):
(WebCore::LibWebRTCMediaEndpoint::stopRTCLogs):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.h:
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCPeerConnectionBackend.cpp:
(WebCore::LibWebRTCPeerConnectionBackend::startGatheringStatLogs):
(WebCore::LibWebRTCPeerConnectionBackend::stopGatheringStatLogs):
(WebCore::LibWebRTCPeerConnectionBackend::provideStatLogs):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCPeerConnectionBackend.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::startGatheringRTCLogs):
(WebCore::Document::stopGatheringRTCLogs):
* Source/WebCore/dom/Document.h:
* Source/WebCore/inspector/CommandLineAPIHost.cpp:
(WebCore::CommandLineAPIHost::gatherRTCLogs):
* Source/WebCore/inspector/CommandLineAPIHost.h:
* Source/WebCore/inspector/CommandLineAPIHost.idl:
* Source/WebCore/inspector/CommandLineAPIModuleSource.js:
(injectModule):
* Source/WebCore/inspector/RTCLogsCallback.h: Added.
* Source/WebCore/inspector/RTCLogsCallback.idl: Added.
* Source/WebCore/platform/SourcesLibWebRTC.txt:
* Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCLogSink.cpp: Added.
(WebCore::LibWebRTCLogSink::LibWebRTCLogSink):
(WebCore::LibWebRTCLogSink::~LibWebRTCLogSink):
(WebCore::toWebRTCLogLevel):
(WebCore::LibWebRTCLogSink::logMessage):
(WebCore::LibWebRTCLogSink::start):
(WebCore::LibWebRTCLogSink::stop):
* Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCLogSink.h: Added.
* Source/WebInspectorUI/UserInterface/Controllers/JavaScriptRuntimeCompletionProvider.js:
(WI.JavaScriptRuntimeCompletionProvider.get _commandLineAPIKeys):

Canonical link: https://commits.webkit.org/274946@main
